blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc

```diff
@@ -0,0 +1,455 @@
+Of course the general idea isn't limited to Hibernate ORM or JPA,
+you could implement something similar with any other API you're using to access your data.
+
+== Creating Aggregate Views via Hibernate ORM
+
+For the following let's assume we're persisting a simple domain model
+(comprising a `Customer` entity and a few related ones like `Address`, (customer) `Category` etc.) in a database.
+Using Hibernate for that allows us to make the creation of aggregates fully transparent to the actual application code using a http://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#events-events[Hibernate event listener].
+Thanks to its extensible architecture, we can plug such listener into Hibernate just by adding it to the classpath,
+from where it will be picked up automatically when bootstrapping the entity manager / session factory.
+
+Our example listener reacts to an annotation, `@MaterializeAggregate`,
+which marks those entity types that should be the roots of materialized aggregates.
+
+[source,java]
+----
+@Entity
+@MaterializeAggregate(aggregateName="customers-complete")
+public class Customer {
+
+    @Id
+    private long id;
+
+    private String firstName;
+
+    @OneToMany(mappedBy = "customer", fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    private Set<Address> addresses;
+
+    @ManyToOne
+    private Category category;
+
+    ...
+}
+----
+
+Now if any entity annotated with `@MaterializeAggregate` is inserted, updated or deleted via Hibernate,
+the listener will kick in and materialize a JSON view of the aggregate root (customer) and its associated entities (addresses, category).
+
+Under the hood the https://github.com/FasterXML/jackson[Jackson API] is used for serializing the model into JSON.
+This means you can use any of its annotations to customize the JSON out, e.g. `@JsonIgnore`  to exclude the inverse relationship from `Address` to `Customer`:
+
+[source,java]
+----
+@Entity
+public class Address {
+
+    @Id
+    private long id;
+
+    @ManyToOne
+    @JoinColumn(name = "customer_id")
+    @JsonIgnore
+    private Customer customer;
+
+    private String street;
+
+    private String city;
+
+    ...
+}
+----
+
+Note that `Address` itself isn't marked with `@MaterializeAggregate`, i.e. it won't be materialized into an aggregate view by itself.
+
+After using JPA's `EntityManager` to insert or update a few customers,
+let's take a look at the `aggregates` table which has been populated by the listener:
+
+[source,sql]
+----
+> select * from aggregates;
+
+| rootType | keySchema | rootId | materialization | valueSchema |
+
+| customers-complete
+
+| {
+  "schema" : {
+    "type" : "struct",
+    "fields" : [ {
+      "type" : "int64",
+      "optional" : false,
+      "field" : "id"
+    } ],
+    "optional" : false,
+    "name" : "customers-complete.Key"
+  }
+}
+
+| { "id" : 1004 }
+
+| { "schema" : { ... } }
```

Comment:
I assume the value schema isn't shown on purpose for brevity reasons?

Reply:
Yes, right. I'll clarify that.